### DISCUSSION TITLE
Board functions findings

### DISCUSSION THREAD

**Discussion by wormstein** [EdNekebno/DGTCentaurMods#14] [Show and tell]
Heya kads,
I leave this here for anyone to decide on how to implemment it till we'll have a config file with all variables we use in the code.
is not a must have feature that's why I didn't put it in the code... but worth knowing. Looks like the same byte is in ledToFrom() for light.

Cheers
```
LED Intensity

\xb0\x00\x0b\x06\x50\x05\x0a\x01\x19\x27\x61
                                  ^   ^   ^--- Checksum
				  |   |
				  |   ------ LED position 
				  |
				  ----------- Intensity byte
								      
x01 - High intensity
x05 - Medium intensity
x19 - Low intensity
```

**Comment by EdNekebno** [comment]
Interesting...I think for now this could be made an optional argument and then specify the possible values and/or put them in variables. At least in boardfunctions.py (as other people might pull the boardfunctions.py file into other stuff and do things it maintains flexibility). 

As an aside, when you write it like that it starts to show some of the pattern in the messages:

First two bytes seem to be the type of functionality it's addressing ( \xb0 = lights. \xb1 = sound. \xb2 = power. \xf0 = boardstate ).
Third byte is the entire message length in bytes. 
Next two are always \x06 \x50 .
Unfortunately then it gets a bit fuzzy until the checksum at the end :( The next byte or two should in theory be the sub-functionality that we want (e.g. turn on one light, turn lights off, etc) and the parameters. But it doesn't seem to be consistent and it might vary by functionality type if it's just sending some raw data to a controller.

**Comment by EdNekebno** [comment]
Oh the rest of the fuzzy unclear bytes do change by function type. e.g. for sound it's just frequency followed by time. e.g. try:

tosend = bytearray(b'\xb1\x00\x08\x06\x50\x4c\x18\x00\x08\x4c\x18\x00\x08\x4c\x18\x00\x08\x4b\x09\x00\x08\x4d\x09\x00\x08\x4c\x18\x00\x08\x4b\x0c\x00\x08\x4d\x09\x00\x08\x4C\x30\x00\x08\x4F\x18\x00\x08\x4F\x18\x00\x08\x4F\x18\x00\x08\x51\x09\x00\x08\x50\x09\x00\x08\x4c\x18\x00\x08\x4a\x09\x00\x08\x4e\x09\x00\x08\x4c\x30\x00\x08\x00');
tosend[2] = len(tosend)
tosend[len(tosend)-1] = boardfunctions.checksum(tosend)
boardfunctions.ser.write(tosend)

or just play this tweet: https://twitter.com/ednekebno/status/1441384529000681475   :)

**Comment by wormstein** [comment]
OMG that's hilarious. :)) We can play a tune at gane end depending on outcome.

I am changing this topic title so we can throw in here any new function we discover.

**Comment by EdNekebno** [comment]
Good idea to collect discoveries here. The lights function follows the same sort of theme when it comes to square numbers - you can have any amount you want as long as the message size byte is correct:

tosend = bytearray(b'\xb0\x00\x0c\x06\x50\x05\x12\x00\x05\x38\x39\x3a\x3b\x3c\x3d\x3e\x3f\x37\x36\x35\x34\x33\x32\x31\x30\x0d')
tosend[2] = len(tosend)
tosend.pop()
tosend.append(boardfunctions.checksum(tosend))
print(tosend.hex())
boardfunctions.ser.write(tosend)

Which breaks down mostly to
\xb0 \x00 = lights
\x0c = message size in bytes
\06 \x50 = unknown but always occurs
\x05 = ??? but assuming it is a lighting subfunction choice. Setting this to 02,03,04 does weird stuff.
\x12 = light flash speed. Higher = faster. Here I set it quite fast. I also tried 0x40 which is very fast but I stopped it and didn't go any higher as I was worried it might damage the board. \x12 is about as high as I'm comfortable setting it.
\x00 = ???? part of above..that would be really really fast :)
\x05 = intensity
\x38..\x30 = the squares to flash. They flash in order. There can be any number of them, just the message size in bytes should match
\x0d = checksum

So the functionality we have in ledFromTo is just a subset of the actual functionality. I guess this is what is used on that centaur startup function that we did the piece detection for that shows with lights which pieces are out of place.

**Reply by EdNekebno** [reply]
Correction to this - the second and third bytes are the message size in bytes. i.e. command, msg size high, msg size low

**Reply by wormstein** [reply]
But looks like the controller ignores as long checksum matches, isn't? I didn't bother to modify any in my tests
 Maybe i was just lucky and had  the same size...

**Reply by EdNekebno** [reply]
Maybe - I never checked. More probably it ignores for some things - for flashing multiple lights for example I guess it will need to know the length properly as there could be multiples. When data comes back from the controller it is also rarely useful. Mostly the entire packet is sent at once. But for some rarer commands you can get multiple packets back.

**Reply by EdNekebno** [reply]
Oh...also it's 7 bit. So to get the message size multiply the first length value by 127 and add the second. Weird.

**Comment by wormstein** [comment]
The controller seems to have a timeout features and it will turn off by itself when idle and leave RPi on. I think I could prevent this by calling a LED once in a while...
It's annoying you must keep play button 10 secs to hard reset it and I think RPi also gets a hard one.

**Reply by EdNekebno** [reply]
When I was playing around early on, the centaur process shutdown the pi a few times. I wonder if the controller sends a message first (like maybe an \xb2 one) and then the centaur process knows to shut down the pi? If it does then menu.py should probably check for that code. I've left mine accidentally over night a couple of times though with ssh connected and in the morning I can carry on just fine (the buttons for using the menu for example still work). Maybe it's something to do with raspbian configuration? Or maybe that I just happen to have the power plugged in. I'll disconnect the power and ssh sessions this morning and see if it times out.

**Reply by wormstein** [reply]
Yup, the shutdown timeout..
 The controller appears to work with the centaur because the behaviour is different if centaur software is running or not. If board is idleing in menu.py, you will find after a while that board is unresponsive but the RPi is still on...

**Comment by EdNekebno** [comment]
The \x06\x50 is a board address.

I worked this out because somehow I was able to toast my board so that this code doesn't work but the original centaur one does. On further checking the command on my board to check for a field changes for example has changed from: \x83\x06\x50\x59  to  \x83\x48\x15\x60 . i.e. I have somehow inadvertently rewritten the address to \x48 \x15 . If I change the address in a test script to get the board to answer a single question it then works.

This is a problem not just for me, but because there could be other boards with different addresses out there. 

There must be some initial communication to get this address which I'll try and reverse. Then every occurrence of \x06\x50 will need to be rewritten with the correct address details :(

**Reply by wormstein** [reply]
This sounds concerning.

There must be a handshake happening in the begining. What about those functions to calibrate the board that we see with pyrasite? Maybe those will just reset the address.

**Reply by EdNekebno** [reply]
Yes...you're right. I have no clue how I changed the address, but there's a handshake at the beginning where the address is passed to the centaur process. I've put it into the beginning of board.py on my local version. The problem is though everywhere that the \x06\x50 was sent it's actually an address and that counts whether it is sent or received - that means I need to adapt practically every single function. Also adapt the checksum for each. Which is bonkers. :( The centaur function does that - which means DGT must've anticipated using different addresses. Very annoying!

**Reply by wormstein** [reply]
Once we figure out the handshake, the easiest way to go will be to have a function we call instead of ser.write where we can manipulate the tosend string. Then we just need to replace ser.write...

I will send you tomorrow a dump of the startup process so you can compare with yours.

**Reply by EdNekebno** [reply]
Locally I think I have the original handshake/passing of the address - I have it like this:

```
addr1 = 0x06
addr2 = 0x50

def checksum(barr):
    csum = 0
    for c in bytes(barr):
        csum += c
    barr_csum = (csum % 128)
    return barr_csum

# But the address might not be that :( Here we send an initial 0x4d to ask the board to provide its address
ser.read(10000)
print("attempt address discovery")
tosend = bytearray(b'\x4d')
ser.write(tosend)
ser.read(10000)
tosend = bytearray(b'\x4e')
ser.write(tosend)
ser.read(10000)
resp = ""
while len(resp) < 4:
    tosend = bytearray(b'\x87\x00\x00\x07')
    ser.write(tosend)
    resp = ser.read(10000)
    if len(resp) > 3:
        addr1 = resp[3]
        addr2 = resp[4]
        print("Discovered new address")
        print(addr1)
        print(addr2)
        print(hex(addr1))
        print(hex(addr2))
        tosend = bytearray(b'\xf4\x00\x07' + addr1.to_bytes(1,byteorder='big') + addr2.to_bytes(1,byteorder='big') + b'\x7f')
        tosend.append(checksum(tosend))
        ser.write(tosend)
        resp = ser.read(10000)
        tosend = bytearray(b'\xf0\x00\x07' + addr1.to_bytes(1,byteorder='big') + addr2.to_bytes(1,byteorder='big') + b'\x7f')
        tosend.append(checksum(tosend))
        ser.write(tosend)
        ser.read(100000)
        resp = "         "
```

That code probably returns \x06 in addr1 and \x50 in addr2 on yours. 

Also when we compare something like a button click it's against the full hex return. Then I have to do something like:

```
                if (resp.hex()[:-2] == "b10011" + "{:02x}".format(addr1) + "{:02x}".format(addr2) + "00140a0501000000007d47"):
                    buttonPress = BTNBACK  # BACK
```

A universal function like you describe would be good, but I think we'd only get as far as a function to tag the checksum on and write it to the serial. Just about everything about this protocol (and I'm learning most chess board protocols in general) is random and weird. The number of bytes before the address varies, as do they afterwards. Sometimes we want to read a response afterwards, occassionally we don't. In a few rare circumstances the address isn't included at all, or a checksum, etc. With the differing possibilities it makes it hard to genericise it very much.

It's not all as bad as I feared though. Luckily \x06\x50 or just 0650 are easy strings to search for and now I have corrected a few it shouldn't be so hard to correct others.

**Reply by EdNekebno** [reply]
Okay...locally I have this fixed I think. I also coded in buildPacket and sendPacket functions to provide a simple way at least of connecting two parts of the data with the address and adding the checksum.

By have it fixed I played a game against an engine at least, I still need to test it with other things. But the code, in principle, will take account of whatever the board says its address is.
